runners/spark/src/main/java/org/apache/beam/runners/spark/io/SparkUnboundedSource.java

```diff
@@ -114,10 +115,16 @@
         .register();
 
     // output the actual (deserialized) stream.
-    WindowedValue.FullWindowedValueCoder<T> coder =
-        WindowedValue.FullWindowedValueCoder.of(
-            source.getDefaultOutputCoder(),
-            GlobalWindow.Coder.INSTANCE);
+    WindowedValue.FullWindowedValueCoder<T> coder;
+
+    try {
+      coder = WindowedValue.FullWindowedValueCoder.of(
+          source.getDefaultOutputCoder(),
```

Comment:
Here too - getDefaultOutputCoder() is allowed to fail and it simply means "the actual coder should be set by the user on the output PCollection" - i.e. if it failed but the user set the coder on the collection, this class should succeed.

In other words, we can not rethrow the exception here, and more like - we must not call source.getDefaultOutputCoder. More like: source.getDefaultOutputCoder() must not be called anywhere except from other methods that are called as part of coder inference; and this call site is one of the violations of this rule.

The fix to this place is to fix spark StreamingTransformTranslator.readUnbounded() to, again, get the coder from the output of the transform (from EvaluationContext) instead of from the source itself.